src/roswire/definitions/package.py

```diff
@@ -82,15 +83,28 @@ def to_dict(self) -> Dict[str, Any]:
 
 class PackageDatabase(Mapping[str, Package]):
     @staticmethod
-    def paths(shell: ShellProxy) -> List[str]:
+    def paths(shell: ShellProxy, file_proxy: FileProxy) -> List[str]:
         """
         Parses the contents of the ROS_PACKAGE_PATH environment variable for a
         given shell.
         """
         code, path_str, duration = shell.execute('echo "${ROS_PACKAGE_PATH}"')
         if code != 0:
             raise Exception("unexpected error when fetching ROS_PACKAGE_PATH")
-        paths: List[str] = path_str.strip().split(':')
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            if file_proxy.isfile(f'{path}/package.xml'):
+                paths.append(path)
+            else:
+                try:
+                    dirs: List[str] = file_proxy.listdir(path, absolute=True)
+                except OSError:
+                    # Not a directory
+                    continue
+                for d in dirs:
+                    if file_proxy.isfile(f'{d}/package.xml'):
```

Comment:
This doesn't work for cases where `package.xml` is nested more than one subdirectory deep. Using a queue is probably better.

Reply:
I tried that. It takes a very long time to traverse all subdirectories. Another way is to just use `find` to list all `package.xml` files. Do you think that's a better solution?

Reply:
That sounds good. It would be great if we could add an interface for `find` to `FileProxy`.

Reply:
I've created a new issue: https://github.com/ChrisTimperley/roswire/issues/181